cpp/src/arrow/compute/function.h

```diff
@@ -229,5 +233,22 @@ class ARROW_EXPORT ScalarAggregateFunction
       const std::vector<ValueDescr>& values) const;
 };
 
+/// \brief A function that dispatches to other functions. Must override
+/// Function::Execute.
+///
+/// For Array, ChunkedArray, and Scalar Datum kinds, may rely on the execution
+/// of concrete Function types, but must handle other Datum kinds on its own.
+class ARROW_EXPORT MetaFunction : public Function {
+ public:
+  int num_kernels() const override { return 0; }
+
+  Result<Datum> Execute(const std::vector<Datum>& args, const FunctionOptions* options,
+                        ExecContext* ctx) const override;
+
+ protected:
```

Comment:
Instead of refusing the bequest by default, I think it'd be better to provide a pure virtual function for MetaFunction subclasses to override:
```suggestion
  Result<Datum> Execute(const std::vector<Datum>& args, const FunctionOptions* options,
                        ExecContext* ctx) const override {
    return ExecuteImpl(args, options, ctx);
  }

 protected:
  virtual Result<Datum> ExecuteImpl(const std::vector<Datum>& args, const FunctionOptions* options,
                        ExecContext* ctx) const = 0;
```

Reply:
done, thank you